Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm

```diff
@@ -1991,3 +1994,354 @@ HTTPServer httpServer({
+
+    // iframe with allowslist="src"
+    const auto iframeWithAllowedPrivateToken = @"iframe = document.createElement(\"iframe\"); iframe.src = \"https://site3.example/path2\"; iframe.allow = \"private-token\"; document.body.appendChild(iframe); true";
+    [webView evaluateJavaScript:iframeWithAllowedPrivateToken completionHandler:^(id value, NSError *error) {
+        EXPECT_NULL(error);
+        EXPECT_WK_STREQ(error.description, @"");
+        doneEvaluatingJavaScript = true;
+    }];
+    Util::run(&doneEvaluatingJavaScript);
+    doneEvaluatingJavaScript = false;
+
+    Util::run(&didReceiveAllowPrivateToken);
+    EXPECT_FALSE(failedNavigation);
+    EXPECT_FALSE(didDecideServerWorkerRequest);
+    EXPECT_TRUE(didReceiveAllowPrivateToken);
+    finishedSuccessfully = false;
+    failedNavigation = false;
+    didDecideServerWorkerRequest = false;
+    didReceiveAllowPrivateToken = false;
+
+    // Result of fetch request initiated by site3.example iframe
+    Util::run(&didReceiveAllowPrivateToken);
+    EXPECT_FALSE(didDecideServerWorkerRequest);
+    EXPECT_TRUE(didReceiveAllowPrivateToken);
+    didReceiveAllowPrivateToken = false;
+
+    // Reload and check only site3.example iframe
+    [webView reload];
+    Util::run(&finishedSuccessfully);
+    EXPECT_FALSE(failedNavigation);
+    EXPECT_FALSE(didDecideServerWorkerRequest);
+    EXPECT_TRUE(didReceiveAllowPrivateToken);
+    finishedSuccessfully = false;
+    failedNavigation = false;
+    didDecideServerWorkerRequest = false;
+    didReceiveAllowPrivateToken = false;
+
+    [webView evaluateJavaScript:iframeWithAllowedPrivateToken completionHandler:^(id value, NSError *error) {
+        EXPECT_NULL(error);
+        EXPECT_WK_STREQ(error.description, @"");
+        doneEvaluatingJavaScript = true;
+    }];
+    Util::run(&doneEvaluatingJavaScript);
+    doneEvaluatingJavaScript = false;
+
+    Util::run(&didReceiveAllowPrivateToken);
+    EXPECT_FALSE(failedNavigation);
+    EXPECT_FALSE(didDecideServerWorkerRequest);
+    EXPECT_TRUE(didReceiveAllowPrivateToken);
+    finishedSuccessfully = false;
+    failedNavigation = false;
+    didDecideServerWorkerRequest = false;
+    didReceiveAllowPrivateToken = false;
+
+    // Result of fetch request initiated by site3.example iframe
+    Util::run(&didReceiveAllowPrivateToken);
+    EXPECT_FALSE(didDecideServerWorkerRequest);
+    EXPECT_TRUE(didReceiveAllowPrivateToken);
+    didReceiveAllowPrivateToken = false;
+
+    // Reload and check only site3.example iframe with a service worker
+    [webView reload];
+    Util::run(&finishedSuccessfully);
+    EXPECT_FALSE(failedNavigation);
+    EXPECT_FALSE(didDecideServerWorkerRequest);
+    EXPECT_TRUE(didReceiveAllowPrivateToken);
+    finishedSuccessfully = false;
+    failedNavigation = false;
+    didDecideServerWorkerRequest = false;
+    didReceiveAllowPrivateToken = false;
+
+    // iframe with allowslist="https://site1.example https://site4.example"
```

Comment:
Maybe we should split those tests as well in their own tests, to ease future maintenance?
This would require introducing a helper routine to share common  test infra code.